src/lib/components/Modal/__tests__/Modal.test.jsx

```diff
@@ -146,27 +155,126 @@ describe('functionality', () => {
   });
 
   it.each([
-    () => userEvent.click(screen.getByText('submit')),
-    () => userEvent.keyboard('{enter}'),
-  ])('do not call submit action when button is disabled (%#)', (action) => {
+    () => userEvent.keyboard('{esc}'),
+    () => userEvent.click(screen.getByTestId('id')),
+  ])('do not call close modal using `closeButtonRef` and `ModalHead` (%#)', (action) => {
     const spy = sinon.spy();
+    const ref = React.createRef();
     render((
       <Modal
-        {...mandatoryProps}
-        actions={[
-          {
-            disabled: true,
-            label: 'submit',
-            onClick: spy,
-            type: 'submit',
-          },
-        ]}
+        closeButtonRef={ref}
+        id="id"
       >
-        content
+        <ModalHead>
+          <ModalCloseButton
+            disabled
+            ref={ref}
+            onClick={spy}
+          />
+        </ModalHead>
       </Modal>
     ));
 
     action();
     expect(spy.called).toEqual(false);
   });
+
+  it('call primary action using `primaryButtonRef`', () => {
+    const spy = sinon.spy();
+    const ref = React.createRef();
+    render((
+      <Modal
+        primaryButtonRef={ref}
+        id="id"
+      >
+        <ModalFooter>
+          <Button
+            label="Submit"
+            onClick={spy}
+            ref={ref}
+            type="button"
+          />
+        </ModalFooter>
+      </Modal>
+    ));
+
+    userEvent.keyboard('{enter}');
+    expect(spy.calledOnce).toEqual(true);
+  });
+
+  it('do not call primary action using `primaryButtonRef ', () => {
+    const spy = sinon.spy();
+    const ref = React.createRef();
+    render((
+      <Modal
+        primaryButtonRef={ref}
+        id="id"
+      >
+        <ModalFooter>
+          <Button
+            disabled
+            label="Submit"
+            onClick={spy}
+            ref={ref}
+            type="button"
+          />
+        </ModalFooter>
+      </Modal>
+    ));
+
+    userEvent.keyboard('{enter}');
+    expect(spy.called).toEqual(false);
+  });
+
+  it.each([
+    [
+      <input />,
+      (rootElement) => {
+        expect(document.activeElement).toEqual(within(rootElement).getByRole('textbox'));
+      },
+    ],
+    [
+      <textarea />,
+      (rootElement) => {
+        expect(document.activeElement).toEqual(within(rootElement).getByRole('textbox'));
+      },
+    ],
+    [
+      (
+        <select>
+          <option>Option</option>
+        </select>
+      ),
+      (rootElement) => {
+        expect(document.activeElement).toEqual(within(rootElement).getByRole('combobox'));
+      },
+    ],
+  ])('autofocuses form field element (%#)', (child, assert) => {
+    const dom = render((
+      <Modal>
+        <ModalBody>
+          {child}
+        </ModalBody>
```

Comment:
I think we should add a primary button here to test that form elements have precedence.